<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузим-данные-и-посмотрим-на-них:" data-toc-modified-id="Загрузим-данные-и-посмотрим-на-них:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузим данные и посмотрим на них:</a></span></li><li><span><a href="#Подготовим-данные" data-toc-modified-id="Подготовим-данные-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подготовим данные</a></span><ul class="toc-item"><li><span><a href="#Стиль-заголовков" data-toc-modified-id="Стиль-заголовков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Стиль заголовков</a></span></li><li><span><a href="#Удалим-избыточные-данные" data-toc-modified-id="Удалим-избыточные-данные-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Удалим избыточные данные</a></span></li><li><span><a href="#Произведём-кодирование-категориальных-признаков-dummy-кодированием" data-toc-modified-id="Произведём-кодирование-категориальных-признаков-dummy-кодированием-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Произведём кодирование категориальных признаков dummy-кодированием</a></span></li></ul></li><li><span><a href="#Разобьём-данные-на-выборки" data-toc-modified-id="Разобьём-данные-на-выборки-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Разобьём данные на выборки</a></span><ul class="toc-item"><li><span><a href="#Отмасштабируем-численные-признаки" data-toc-modified-id="Отмасштабируем-численные-признаки-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Отмасштабируем численные признаки</a></span></li><li><span><a href="#Исследуем-баланс-классов" data-toc-modified-id="Исследуем-баланс-классов-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Исследуем баланс классов</a></span></li></ul></li><li><span><a href="#Исследуем-модель-без-учета-дисбаланса-классов" data-toc-modified-id="Исследуем-модель-без-учета-дисбаланса-классов-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Исследуем модель без учета дисбаланса классов</a></span></li><li><span><a href="#Исследуем-модель-с-учетом-дисбаланса-классов" data-toc-modified-id="Исследуем-модель-с-учетом-дисбаланса-классов-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Исследуем модель с учетом дисбаланса классов</a></span><ul class="toc-item"><li><span><a href="#Взвешивание-классов" data-toc-modified-id="Взвешивание-классов-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Взвешивание классов</a></span></li><li><span><a href="#Увеличение-и-уменьшение-выборки" data-toc-modified-id="Увеличение-и-уменьшение-выборки-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Увеличение и уменьшение выборки</a></span></li><li><span><a href="#Решающее-дерево" data-toc-modified-id="Решающее-дерево-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Решающее дерево</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Выводы</a></span></li><li><span><a href="#Проверим-модели-на-тестовой-выборке" data-toc-modified-id="Проверим-модели-на-тестовой-выборке-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Проверим модели на тестовой выборке</a></span></li></ul></div>

# Исследование ухода клиентов

Из банка каждый месяц по-немногу стали уходить клиенты. Сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно построить модель для прогноза, уйдёт клиент из банка в ближайшее время или нет. Модель должна быть с предельно большим значением F1-меры  (0.59 и выше). Дополнительно измеряем AUC-ROC, сравниваем её значение с F1-мерой.


Исторические данные о поведении клиентов и расторжении договоров с банком из источника данных: https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling

In [70]:
import pandas as pd
from sklearn.model_selection import train_test_split # чтобы делить выборки
from sklearn.preprocessing import StandardScaler # чтобы отмасштабировать значения
from sklearn.utils import shuffle # чтобы перемешивать объекты в выборке
from sklearn.linear_model import LogisticRegression # алгоритм логистической регрессии
from sklearn.tree import DecisionTreeClassifier # чтобы обучать модель решающего дерева для классификации
from sklearn.ensemble import RandomForestClassifier # алгоритм случайного леса
# чтобы оценивать качество моделей
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score

import warnings
warnings.filterwarnings('ignore')

## Загрузим данные и посмотрим на них:

In [71]:
df = pd.read_csv('/datasets/Churn.csv', index_col='RowNumber')

Посмотрим на первые десять строк таблицы:

In [72]:
df.head(10)

CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
RowNumber                                                                     
1            15634602  Hargrave          619    France  Female   42     2.0   
2            15647311      Hill          608     Spain  Female   41     1.0   
3            15619304      Onio          502    France  Female   42     8.0   
4            15701354      Boni          699    France  Female   39     1.0   
5            15737888  Mitchell          850     Spain  Female   43     2.0   
6            15574012       Chu          645     Spain    Male   44     8.0   
7            15592531  Bartlett          822    France    Male   50     7.0   
8            15656148    Obinna          376   Germany  Female   29     4.0   
9            15792365        He          501    France    Male   44     4.0   
10           15592389        H?          684    France    Male   27     2.0   

             Balance  NumOfProducts  HasCrCard  IsActiveMember  \
RowNumber                                                        
1               0.00              1          1               1   
2           83807.86              1          0               1   
3          159660.80              3          1               0   
4               0.00              2          0               0   
5          125510.82              1          1               1   
6          113755.78              2          1               0   
7               0.00              2          1               1   
8          115046.74              4          1               0   
9          142051.07              2          0               1   
10         134603.88              1          1               1   

           EstimatedSalary  Exited  
RowNumber                           
1                101348.88       1  
2                112542.58       0  
3                113931.57       1  
4                 93826.63       0  
5                 79084.10       0  
6                149756.71       1  
7                 10062.80       0  
8                119346.88       1  
9                 74940.50       0  
10                71725.73       0

Каждый объект в наборе данных — это информация о поведении одного клиента. Известно:

- `RowNumber` — индекс строки в данных
-	`CustomerId` — уникальный идентификатор клиента
-	`Surname` — фамилия
-	`CreditScore` — кредитный рейтинг
-	`Geography` — страна проживания
-	`Gender` — пол
-	`Age` — возраст
-	`Tenure` — сколько лет человек является клиентом банка
-	`Balance` — баланс на счёте
-	`NumOfProducts` — количество продуктов банка, используемых клиентом
-	`HasCrCard` — наличие кредитной карты (1 - есть)
-	`IsActiveMember` — активность клиента (1 - активен)
-	`EstimatedSalary` — предполагаемая зарплата

Целевой признак
-	`Exited` — факт ухода клиента (1 - ушел)


## Подготовим данные


### Стиль заголовков
Выведем на экран названия столбцов:


In [73]:
df.columns

Index(['CustomerId', 'Surname', 'CreditScore', 'Geography', 'Gender', 'Age',
       'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

Переменуем столбцы согласно хорошему стилю:

In [74]:
df= df.rename(columns = {'CustomerId':'customer_id','Surname':'surname','CreditScore':'credit_score',
                         'Geography':'geography' ,'Gender':'gender','Age':'age',
                         'Tenure':'tenure','Balance':'balance',
                         'NumOfProducts':'num_of_products' ,'HasCrCard':'has_cr_card','IsActiveMember':'is_active_member',
                         'EstimatedSalary':'estimated_salary', 'Exited':'exited'})

In [75]:
df.columns

Index(['customer_id', 'surname', 'credit_score', 'geography', 'gender', 'age',
       'tenure', 'balance', 'num_of_products', 'has_cr_card',
       'is_active_member', 'estimated_salary', 'exited'],
      dtype='object')


### Удалим избыточные данные


In [76]:
df.duplicated().sum()

0

Явных дубликатов в данных нет. Удалим столбец с фамилиями и уникальными индификаторами клиентов: модели эти данные никак не помогут, их нельзя выразить численно

In [77]:
df = df.drop(['customer_id','surname'], axis=1)

In [78]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 1 to 10000
Data columns (total 11 columns):
credit_score        10000 non-null int64
geography           10000 non-null object
gender              10000 non-null object
age                 10000 non-null int64
tenure              9091 non-null float64
balance             10000 non-null float64
num_of_products     10000 non-null int64
has_cr_card         10000 non-null int64
is_active_member    10000 non-null int64
estimated_salary    10000 non-null float64
exited              10000 non-null int64
dtypes: float64(3), int64(6), object(2)
memory usage: 937.5+ KB


В атрибуте "сколько лет человек является клиентом банка" есть пропуски, заменим их медианными значениями согласно возрасту клиента.

In [79]:
#Создадим новый столбец, в котором указаны медианные значения количества лет сотрудничества для каждой строки согласно возрасту клиента
median_tenure = df.groupby('age')['tenure'].transform('median')

In [80]:
# заполним все пустые значения количества лет сотрудничества на медианные значения из нового столбца
df['tenure'] = df['tenure'].fillna(median_tenure)

### Произведём кодирование категориальных признаков dummy-кодированием

A именно страну и пол клиентов.

In [81]:
df['geography'].value_counts(dropna=False)

France     5014
Germany    2509
Spain      2477
Name: geography, dtype: int64

In [82]:
df_ohe = pd.get_dummies(df, drop_first=True)

Применим масштабирование на обучающие, валидационные и тестовые признаки в следующем разделе:

## Разобьём данные на выборки

Все объекты из датафрейма необходимо разделить на три выборки: тренировочную, валидационную и тестовую.

На тренировочной модели мы будем обучать модель, на вадидационной проверять работоспособность модели (выявлять переобучение, например), а на тестовой её тестировать. Разобъём данные на выборки в соотношении 3:1:1.

Первым шагом разобъём данные на целевой признак (ушел ли человек из банка) и признаки, которые подвели человека к уходу/не уходу.

In [84]:
features = df_ohe.drop(['exited'], axis=1)

In [85]:
target = df_ohe['exited'].copy()

Разделим данные на две выборки: данные, необходимые до тестирования и данные для тестов. Потом данные, нужные до тестирования разделим на тренировочные и валидационные.

In [86]:
 # отделим 20% данных для тестовой выборки
features_before_test, features_test, target_before_test, target_test = train_test_split(features, target, test_size=0.2, random_state=12345).copy()

In [87]:
 # отделим 25% данных для валидационной выборки
features_train, features_valid, target_train, target_valid = train_test_split(features_before_test, target_before_test, test_size=0.25, random_state=12345).copy()

In [88]:
features_train.shape

(6000, 11)

In [89]:
features_valid.shape

(2000, 11)

In [90]:
target_train.shape

(6000,)

In [91]:
target_valid.shape

(2000,)

In [92]:
features_test.shape

(2000, 11)

In [93]:
target_test.shape

(2000,)

### Отмасштабируем численные признаки

In [94]:
numeric = ['credit_score', 'age', 'tenure', 'balance', 'num_of_products', 'estimated_salary']

scaler = StandardScaler()
scaler.fit(features_train[numeric]) 

StandardScaler(copy=True, with_mean=True, with_std=True)

Применим масштабирование на обучающие, валидационные и тестовые признаки:

In [95]:
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric]) 
features_test[numeric] = scaler.transform(features_test[numeric])

Разделили объекты на три выборки, положили их в отдельные переменные:
- features_train, target_train это выборки для тренировки модели,
-  features_valid, target_valid - выборки для валидации модели,
- features_test, target_test - выборки для тестирования.

Отмасштабировали численные признаки, чтобы модель училась более корректно.

### Исследуем баланс классов

In [96]:
df['exited'].mean()

0.2037

In [97]:
df['exited'].count()

10000

20% клиентов из 10000 клиентов датасета ушли из банка. Остальные продолжают пользоваться услугами банка.

Можно сказать, что в этой выборке есть дисбаланс классов. Обучим модель без учёта дисбаланса, а потом учитывая дисбаланс классов.

## Исследуем модель без учета дисбаланса классов






Обучим модель на основе решающего дерева. Посмотрим различную глубину дерева, выберем оптимальную модель.

In [98]:
best_model_tree = None
best_result = 0
best_depth = None
for depth in range(1, 12):
  model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
  model.fit(features_train, target_train) # обучим модель

  train_predictions = model.predict(features_train) # получим предсказания модели
  valid_predictions = model.predict(features_valid) 
  
  f1_train = f1_score(target_train,train_predictions) # посчитаем качество модели f1
  f1_valid = f1_score(target_valid,valid_predictions)


  print (f'Учебная выборка: {f1_train:.5f}, валидационная выборка: {f1_valid:.5f}, глубина дерева: {depth}')
  if f1_valid > best_result:
    best_model_tree = model
    best_result = f1_valid
    best_depth = depth
        
print("Лучшая глубина дерева", best_depth)

probabilities_valid = best_model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
print(f"Для этого дерева AUC ROC {auc_roc:.5f} а f1 {best_result:.5f}")

Учебная выборка: 0.00000, валидационная выборка: 0.00000, глубина дерева: 1
Учебная выборка: 0.51013, валидационная выборка: 0.50370, глубина дерева: 2
Учебная выборка: 0.41469, валидационная выборка: 0.39382, глубина дерева: 3
Учебная выборка: 0.45216, валидационная выборка: 0.43019, глубина дерева: 4
Учебная выборка: 0.58737, валидационная выборка: 0.54884, глубина дерева: 5
Учебная выборка: 0.56233, валидационная выборка: 0.51130, глубина дерева: 6
Учебная выборка: 0.62575, валидационная выборка: 0.55836, глубина дерева: 7
Учебная выборка: 0.67676, валидационная выборка: 0.53988, глубина дерева: 8
Учебная выборка: 0.71159, валидационная выборка: 0.53571, глубина дерева: 9
Учебная выборка: 0.74728, валидационная выборка: 0.53835, глубина дерева: 10
Учебная выборка: 0.78351, валидационная выборка: 0.51320, глубина дерева: 11
Лучшая глубина дерева 7
Для этого дерева AUC ROC 0.82310 а f1 0.55836


Чтобы выявить, как сильно наша модель отличается от случайной смотрим на параметр AUC ROC. Если он около 0.5, значит наша модель не сильно отличается от случайной. Хорошо, когда этот параметр стремится к единице.


F1-мера - это среднее гармоническое полноты и точности. Единица в F1 означает, что соотношение полноты и точности равно 1:1. По отдельности полнота и точность не слишком информативны. 
Полнота и точность оценивают качество прогноза положительного класса с разных позиций. Recall(полнота) описывает, как хорошо модель разобралась в особенностях этого класса и распознала его. Precision(точность) выявляет, не переусердствует ли модель, присваивая положительные метки. 
Важны обе метрики. Контролировать их параллельно помогают агрегирующие метрики, одна из которых — F1-мера. 

Обученная модель ещё далека от идеала: f1 метрику необходимо довести хотябы до 0.59.

Попробуем обучить модели с учетом дибаланса классов:

## Исследуем модель с учетом дисбаланса классов

### Взвешивание классов

Необходимо при обучении моделей указывать `class_weight='balanced'`

Если указать class_weight='balanced', алгоритм посчитает, во сколько раз класс «0» встречается чаще класса «1» у целевого признака. 
Бóльший вес будет у редкого класса.

### Увеличение и уменьшение выборки

Сделаем объекты редкого класса не такими редкими в данных.

Применим технику upsampling, преобразование проходит в несколько этапов:

* Разделим обучающую выборку на отрицательные и положительные объекты;
* Скопируем несколько раз положительные объекты;
* С учётом полученных данных создадим новую обучающую выборку;
* Перемешаем данные.

Также сделаем объекты частого класса не такими частыми.
Применим технику downsampling, преобразование проходит в несколько этапов:

* Разделим обучающую выборку на отрицательные и положительные объекты;
* Случайным образом отбросим часть из отрицательных объектов;
* С учётом полученных данных создадим новую обучающую выборку;
* Перемешаем данные.

Сбалансируем данные:
напишем функцию для балансировки.

Сбалансируем только тренировочные выборки.

In [99]:
def up_down_sample(features, target, repeat, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    # Уменьшаем отрицательные признаки в выборке
    features_sample = features_zeros.sample(frac=fraction, random_state=12345)
    target_sample = target_zeros.sample(frac=fraction, random_state=12345)

    # создаём новую выборку  из уменьшенных отрицательных признаков и увеличенных положительных признаков
    features_updownsampled = pd.concat([features_sample] + [features_ones] * repeat)
    target_updownsampled = pd.concat([target_sample] + [target_ones] * repeat)
    
    # перемешиваем потрицательные и положительные признаки
    features_updownsampled, target_updownsampled = shuffle(
        features_updownsampled, target_updownsampled, random_state=12345)
    
    return features_updownsampled, target_updownsampled

In [100]:
features_update_train, target_update_train = up_down_sample(features_train, target_train, 2, 0.9)

Теперь для обучения моделей будем использовать выборки `features_update_train` и `target_update_train`.

Также попробуем использовать взвешивание классов.

### Решающее дерево

Построим и обучим модель на основе решающего дерева. 

Будем смотреть на метрики f1 и AUC ROC.


Чтобы получить максимально высокое качество предсказания нужно перебрать в алгоритме обучения разные варианты глубины дерева.
Найдём модель с самой большой долей правильных ответов. Будем рассматривать глубину дерева от 1 до 11.

In [101]:
best_model_tree = None
best_result = 0
best_depth = None
for depth in range(1, 12):
  model = DecisionTreeClassifier(random_state=12345, max_depth=depth, class_weight='balanced') 
  model.fit(features_update_train, target_update_train) # обучим модель на новых данных

  train_predictions = model.predict(features_update_train) # получим предсказания модели
  valid_predictions = model.predict(features_valid) 
  
  f1_train = f1_score(target_update_train,train_predictions) # посчитаем качество модели f1
  f1_valid = f1_score(target_valid,valid_predictions)


  print (f'Учебная выборка: {f1_train:.5f}, валидационная выборка: {f1_valid:.5f}, глубина дерева: {depth}')
  if f1_valid > best_result:
    best_model_tree = model
    best_result = f1_valid
    best_depth = depth
        
print("Лучшая глубина дерева", best_depth)

probabilities_valid = best_model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
print(f"Для этого дерева AUC ROC {auc_roc:.5f} а f1 {best_result:.5f}")

Учебная выборка: 0.60758, валидационная выборка: 0.47507, глубина дерева: 1
Учебная выборка: 0.64003, валидационная выборка: 0.49762, глубина дерева: 2
Учебная выборка: 0.64003, валидационная выборка: 0.49762, глубина дерева: 3
Учебная выборка: 0.68436, валидационная выборка: 0.51656, глубина дерева: 4
Учебная выборка: 0.71061, валидационная выборка: 0.54201, глубина дерева: 5
Учебная выборка: 0.73116, валидационная выборка: 0.55945, глубина дерева: 6
Учебная выборка: 0.74860, валидационная выборка: 0.55061, глубина дерева: 7
Учебная выборка: 0.77502, валидационная выборка: 0.56897, глубина дерева: 8
Учебная выборка: 0.80047, валидационная выборка: 0.53019, глубина дерева: 9
Учебная выборка: 0.83260, валидационная выборка: 0.49307, глубина дерева: 10
Учебная выборка: 0.86219, валидационная выборка: 0.48665, глубина дерева: 11
Лучшая глубина дерева 8
Для этого дерева AUC ROC 0.80135 а f1 0.56897


Модель переобучилась, поскольку F1 этой модели на тренировочной выборке больше чем на валидационной.

### Случайный лес

Попробуем другой алгоритм классификации — случайный лес. Алгоритм обучает большое количество независимых друг от друга деревьев, а потом принимает решение на основе голосования. Случайный лес помогает улучшить результат предсказания и избежать переобучения.

In [102]:
best_model_forest = None
best_result = 0
best_est = 0
best_depth = 0
for est in range(10, 51, 10):
  for depth in range (1, 11):
    model = RandomForestClassifier(random_state=12345, n_estimators=est, max_depth=depth, class_weight='balanced') # создадим модель с заданным количеством деревьев и глубиной дерева
    model.fit(features_update_train, target_update_train) # обучим модель на тренировочной выборке
    valid_predictions = model.predict(features_valid) 
    f1_valid = f1_score(target_valid,valid_predictions) # посчитаем качество модели на валидационной выборке
    if f1_valid > best_result:
        best_model_forest = model # сохраним наилучшую модель
        best_result = f1_valid #  сохраним наилучшее значение метрики f1 на валидационных данных
        best_est = est
        best_depth = depth

probabilities_valid = best_model_forest.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

print(f"F1 наилучшей модели на валидационной выборке: {best_result:.5f}, AUC ROC {auc_roc:.5f}. Количество деревьев: {best_est}, Максимальная глубина: {best_depth}")


F1 наилучшей модели на валидационной выборке: 0.60023, AUC ROC 0.84787. Количество деревьев: 50, Максимальная глубина: 10


### Логистическая регрессия

Алгоритмы не ограничиваются деревьями. Есть и другие способы классификации.

инициализируем модель логистической регрессии с параметром random_state=12345

In [103]:
model_log_reg = LogisticRegression(random_state=12345, class_weight='balanced') 

 Обучим модель на тренировочной выборке

In [104]:
model_log_reg.fit(features_update_train, target_update_train)

LogisticRegression(C=1.0, class_weight='balanced', dual=False,
                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,
                   max_iter=100, multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=12345, solver='warn', tol=0.0001, verbose=0,
                   warm_start=False)

Получим метрику F1 модели на валидационной выборке:


In [105]:
valid_predictions = model_log_reg.predict(features_valid) 
f1_valid = f1_score(target_valid,valid_predictions)
f1_valid 

0.4755120213713268

И метрику AUC ROC:

In [106]:
probabilities_valid = model_log_reg.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]
auc_roc = roc_auc_score(target_valid, probabilities_one_valid)
auc_roc

0.7724246128315948

Получилось не очень много, зато быстро.

## Выводы




Оценим модели по качеству и скорости работы.
1. Самое высокое качество у случайного леса: вместо одного решающего дерева используется целый лес.
На втором месте — дерево решений. При глубине дерева равной 8 модель дала хороший (но не достаточный) результат.
Самое низкое качество предсказания у логистической регрессии, с этой задачей она справилась плохо.
2. Скорость работы тоже значимый критерий.
Высокая скорость работы у логистической регрессии: у неё меньше всего параметров.
Скорость решающего дерева тоже высокая и зависит от глубины. Наша модель получила ответ на вопрос за 8 проверок значений признаков — это быстро. Случайный лес медленнее всех: чем больше деревьев, тем медленнее работает модель.

F1 наилучшей модели 0.60, AUC ROC 0.848. Для случайного леса. Количество деревьев: 50, Максимальная глубина: 10

Для дерева решений глубиной 8 F1 = 0.57, а AUC ROC = 0.8.

Метрики логистической регрессии и дерева решений не дотянули до необходимого порога.


## Проверим модели на тестовой выборке





Протестируем:

In [110]:
test_predictions  = best_model_tree.predict(features_test)

f1_test = f1_score(target_test,test_predictions)

probabilities_test = best_model_tree.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc = roc_auc_score(target_test, probabilities_one_test)

print(f"Для этого дерева f1 {f1_test:.5f}, a AUC ROC {auc_roc:.5f} ")

Для этого дерева f1 0.60545, a AUC ROC 0.81001 


In [111]:
test_predictions = best_model_forest.predict(features_test) 
f1_valid = f1_score(target_test,test_predictions) # посчитаем качество модели на валидационной выборке

probabilities_test = best_model_forest.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc = roc_auc_score(target_test, probabilities_one_test)

print(f"Для этого леса f1 {f1_test:.5f}, a AUC ROC {auc_roc:.5f} ")

Для этого леса f1 0.60545, a AUC ROC 0.86196 


In [112]:
test_predictions = model_log_reg.predict(features_test) 
f1_valid = f1_score(target_test, test_predictions)

probabilities_test = model_log_reg.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]
auc_roc = roc_auc_score(target_test, probabilities_one_test)

print(f"Для логистической регрессии f1 {f1_test:.5f}, a AUC ROC {auc_roc:.5f} ")

Для логистической регрессии f1 0.60545, a AUC ROC 0.76340 


Внезапно, у всех моделей на тестовой выборке значение F1 больше 0,6!

Поскольку случайный лес показал себя лучше на валидационной выборке, то предпочтительно выбрать его для предсказания оттока клиентов.